iree/compiler/Conversion/Common/LinalgBufferizePass.cpp

```diff
@@ -469,16 +438,33 @@ static LogicalResult analyseSubTensorOp(SubTensorOp subTensorOp,
 static LogicalResult analyseDestructiveUpdateOp(Operation *op, Value source,
                                                 Value dest, Value result,
                                                 BufferizationPlan &plan) {
-  if (dest.hasOneUse() && !isFromReadOnlyTensor(dest)) {
+  if (hasSingleRealUse(dest) && !isFromReadOnlyTensor(dest, plan)) {
     plan.unionSets(dest, result);
   } else if (source && plan.isEquivalent(source, dest)) {
-    return success();
+    // The destructive update pattern can put the source and dest in the same
+    // equivalence class, but that is checked explicitly later on. So at this
+    // stage this shouldnt happen.
+    return op->emitError(
+        "unexpected source and dest being equivalent in destructive update op");
```

Comment:
If this should not happen at this stage, do we put an assertion instead?

Reply:
Assertion will only happen with the right builds. I'd just prefer an error.